File: python/aoc_2023/day07.py
import enum
from dataclasses import dataclass
from typing import List, Tuple


class Card(enum.Enum):
    "AKQJT98765432"
    TWO = 2
    THREE = 3
    FOUR = 4
    FIVE = 5
    SIX = 6
    SEVEN = 7
    EIGHT = 8
    NINE = 9
    TEN = 10
    JACK = 11
    QUEEN = 12
    KING = 13
    ACE = 14
    JOKER = 1

    @staticmethod
    def from_string(s: str):
        mapping = {
            "2": Card.TWO,
            "3": Card.THREE,
            "4": Card.FOUR,
            "5": Card.FIVE,
            "6": Card.SIX,
            "7": Card.SEVEN,
            "8": Card.EIGHT,
            "9": Card.NINE,
            "T": Card.TEN,
            "J": Card.JACK,
            "Q": Card.QUEEN,
            "K": Card.KING,
            "A": Card.ACE,
            "X": Card.JOKER,
        }
        return mapping[s]

    
class HandType(enum.Enum):
    HIGH_CARD = 1
    ONE_PAIR = 2
    TWO_PAIR = 3
    THREE_OF_A_KIND = 4
    FULL_HOUSE = 5
    FOUR_OF_A_KIND = 6
    FIVE_OF_A_KIND = 7
    
    @staticmethod
    def from_cards(cards: List[Card]):
        """
        Jokers are a special case here. They can be a match to any other card for the purpose of
        making the highest value HandType.
        """
        assert len(cards) == 5
        
        counts = {}
        for card in cards:
            if card not in counts:
                counts[card] = 0
            counts[card] += 1

        # For any jokers, add them to the count of the most common card
        if Card.JOKER in counts:
            most_common = None
            for card, count in counts.items():
                if most_common is None or count > counts[most_common]:
                    most_common = card
            counts[most_common] += counts[Card.JOKER]

        # Five of a kind
        if 5 in counts.values():
            return HandType.FIVE_OF_A_KIND
        
        # Four of a kind
        if 4 in counts.values():
            return HandType.FOUR_OF_A_KIND
        
        # Full house
        if 3 in counts.values() and 2 in counts.values():
            return HandType.FULL_HOUSE
        
        # Three of a kind
        if 3 in counts.values():
            return HandType.THREE_OF_A_KIND
        
        # Two pair
        if list(counts.values()).count(2) == 2:
            return HandType.TWO_PAIR
        
        # One pair
        if 2 in counts.values():
            return HandType.ONE_PAIR
        
        # High card
        return HandType.HIGH_CARD


        

       


        
        

@dataclass
class Hand:
    cards: List[Card]
    hand_type: HandType
    value: Tuple[int]
    bid: int
    
    @staticmethod
    def from_string(s: str):
        "Line looks like: 32T3K 765 (cards and bid)"
        card_symbols, bid = s.split()
        cards = [Card.from_string(c) for c in card_symbols]
        hand_type = HandType.from_cards(cards)
        value = (hand_type.value, *(c.value for c in cards))
        return Hand(cards, hand_type, value, int(bid))
    
    def __lt__(self, other: 'Hand'):
        return self.value < other.value
    
    
    

File: rust/src/day07.rs
use std::cmp::Ordering;
use std::collections::HashMap;

#[derive(Debug, PartialEq, Eq, Hash, Clone, Copy)]
enum Card {
    Two = 2,
    Three = 3,
    Four = 4,
    Five = 5,
    Six = 6,
    Seven = 7,
    Eight = 8,
    Nine = 9,
    Ten = 10,
    Jack = 11,
    Queen = 12,
    King = 13,
    Ace = 14,
    Joker = 1,
}

impl Card {
    fn from_char(c: char) -> Card {
        match c {
            '2' => Card::Two,
            '3' => Card::Three,
            '4' => Card::Four,
            '5' => Card::Five,
            '6' => Card::Six,
            '7' => Card::Seven,
            '8' => Card::Eight,
            '9' => Card::Nine,
            'T' => Card::Ten,
            'J' => Card::Jack,
            'Q' => Card::Queen,
            'K' => Card::King,
            'A' => Card::Ace,
            'X' => Card::Joker,
            _ => panic!("Invalid card: {}", c),
        }
    }
}

#[derive(Debug, PartialEq, Eq, Clone, Copy)]
enum HandType {
    HighCard = 1,
    OnePair = 2,
    TwoPair = 3,
    ThreeOfAKind = 4,
    FullHouse = 5,
    FourOfAKind = 6,
    FiveOfAKind = 7,
}

impl HandType {
    fn from_cards(cards: &[Card]) -> HandType {
        let mut counts = HashMap::new();
        for &card in cards {
            *counts.entry(card).or_insert(0) += 1;
        }

        // Special handling for jokers
        if let Some(&joker_count) = counts.get(&Card::Joker) {
            counts.remove(&Card::Joker);
            if joker_count == 5 {
                return HandType::FiveOfAKind;
            }

            if let Some((most_common_card, _)) = counts.iter().max_by_key(|&(_, count)| count) {
                *counts.entry(*most_common_card).or_insert(0) += joker_count;
            }
        }

        match *counts.values().max().unwrap_or(&0) {
            5 => HandType::FiveOfAKind,
            4 => HandType::FourOfAKind,
            3 if counts.values().any(|&v| v == 2) => HandType::FullHouse,
            3 => HandType::ThreeOfAKind,
            2 if counts.values().filter(|&&v| v == 2).count() == 2 => HandType::TwoPair,
            2 => HandType::OnePair,
            _ => HandType::HighCard,
        }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct Hand {
    cards: Vec<Card>,
    hand_type: HandType,
    value: (usize, Vec<u8>),
    pub bid: usize,
}

impl Hand {
    pub fn from_string(s: String) -> Option<Hand> {
        let mut parts = s.split_whitespace();
        let card_symbols = parts.next()?;
        let bid = parts.next()?.parse().ok()?;

        let cards: Vec<Card> = card_symbols.chars().map(|c| Card::from_char(c)).collect();

        if cards.len() != 5 {
            println!("Invalid hand: {:?}", cards);
            return None;
        }

        let hand_type = HandType::from_cards(&cards);
        let value = (hand_type as usize, cards.iter().map(|&c| c as u8).collect());

        Some(Hand {
            cards,
            hand_type,
            value,
            bid,
        })
    }
}

impl PartialOrd for Hand {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.value.cmp(&other.value))
    }
}

impl Ord for Hand {
    fn cmp(&self, other: &Self) -> Ordering {
        self.value.cmp(&other.value)
    }
}
